test/unit/settings_test.py

```diff
@@ -126,20 +126,17 @@ def test_none_quarters_raise(self: Self):
 
 
 class TestEIA860Settings:
-    """Test EIA860 setting validation.
-
-    Most of the validation is covered in TestFerc1Settings.
-    """
+    """Test EIA860 setting validation."""
 
     def test_860m(self: Self):
         """Test validation error is raised when eia860m date is within 860 years."""
-        settings_cls = Eia860Settings
-        original_eia80m_year_month = settings_cls.eia860m_year_month
-        settings_cls.eia860m_year_month = "2019-11"
+        settings_cls = Eia860Settings()
+        original_eia80m_year_month = settings_cls.eia860m_year_months
+        settings_cls.eia860m_year_months = ["2019-11"]
 
-        with pytest.raises(ValidationError):
+        with pytest.raises(ValueError):
             settings_cls(eia860m=True)
-        settings_cls.eia860m_year_month = original_eia80m_year_month
+        settings_cls.eia860m_year_months = original_eia80m_year_month
```

Comment:
For some reason when I run this test `pytest test/unit/settings_test.py::TestEIA860Settings::test_860m` I get the following error: 

```
name = 'eia860m_year_months', value = ['2019-11']

    def _check_frozen(self, name: str, value: Any) -> None:
        if self.model_config.get('frozen', None):
            typ = 'frozen_instance'
        elif getattr(self.model_fields.get(name), 'frozen', False):
            typ = 'frozen_field'
        else:
            return
        error: pydantic_core.InitErrorDetails = {
            'type': typ,
            'loc': (name,),
            'input': value,
        }
>       raise pydantic_core.ValidationError.from_exception_data(self.__class__.__name__, [error])
E       pydantic_core._pydantic_core.ValidationError: 1 validation error for Eia860Settings
E       eia860m_year_months
E         Instance is frozen [type=frozen_instance, input_value=['2019-11'], input_type=list]
E           For further information visit https://errors.pydantic.dev/2.7/v/frozen_instance

../../../mambaforge/envs/pudl-dev/lib/python3.12/site-packages/pydantic/main.py:894: ValidationError
```

I don't know why this error wasn't happening before -- the test was still replacing the class variable `eia860m_year_month`. The only difference was that it was a single string value rather than a list of strings.

Reply:
Oh also, I changed it from a `Validation` to a `Value` error because that's what error the `only_years_not_in_eia860` function should throw. 

...on a separate note, I'd love to better understand the difference between validation and value errors

Reply:
`ValidationError` is [defined by Pydantic](https://docs.pydantic.dev/latest/errors/validation_errors/), while `ValueError` is [a built-in error type](https://docs.python.org/3/library/exceptions.html#ValueError) that's part of Python:

> Raised when an operation or function receives an argument that has the right type but an inappropriate value

Pydantic gathers together information from all of the different validation failures and tries to present it together for use in debugging. [Pydantic says](https://docs.pydantic.dev/latest/concepts/validators/#handling-errors-in-validators) validators should throw either `ValueError` or `AssertionError` when they fail, so I assume the validation machinery is specifically trying to catch those exceptions in compiling the more extensive `ValidationError` output.

Reply:
@aesharpe  I will test this locally but I believe this is happening because you are instantiating the object in this first line (`settings_cls = Eia860Settings()`) when it was previously `settings_cls = Eia860Settings`

Reply:
but if you don't instantiate the class you no longer get access to the default `eia860m_year_months`. This attribute was a `ClassVar` it was accessible w/o instantiating the class. But you can still grab the default value is you just instantiate a version of the class. i thiiink to get this tests as is to run properly you'll need to do something like this:

```python
settings_cls = Eia860Settings
original_eia80m_year_month = Eia860Settings().eia860m_year_months
settings_cls.eia860m_year_months = ["2019-11"]

with pytest.raises(ValueError):
    settings_cls(eia860m=True)
settings_cls.eia860m_year_months = original_eia80m_year_month
```

buuuutttt this test is testing the previous classes functionality. its testing the previous field validation of the `eia860m` boolean. Because you have added several good new validations, it would be good to add lil tests to those failure modes.

Reply:
I would add something like this:

```python
class TestEIA860Settings:
    """Test EIA860 setting validation."""
    def test_860m_year_is_one_more_than_860_max_year(self: Self):
        """Test whether or not the first EIA860m year is the next year after the max 860 year."""
        settings_eia860= Eia860Settings(eia860m_year_months=["2024-01"])
        assert max(settings_eia860.years)+1 == int(min(settings_eia860.eia860m_year_months).split("-")[0])
```

Reply:
also something weird is happening bc:

```python
settings_eia860= Eia860Settings(years=[2022], eia860m_year_months=["2019-01"])
min(pd.to_datetime(settings_eia860.eia860m_year_months).year)
>> 2015
```

Reply:
> ```python
> settings_cls = Eia860Settings
> original_eia80m_year_month = Eia860Settings().eia860m_year_months
> settings_cls.eia860m_year_months = ["2019-11"]
> 
> with pytest.raises(ValueError):
>     settings_cls(eia860m=True)
> settings_cls.eia860m_year_months = original_eia80m_year_month
> ```

I made this suggested edit, and it doesn't seem to be working. The test fails because it doesn't raise a `ValueError`. I added a `breakpoint()` to the `only_years_not_in_eia860()` function in `Eia860Settings` that it should be testing, and it doesn't stop when I run the tests. This leads me to believe that the `only_years_not_in_eia860` function isn't getting called in the test?

Reply:
I think the problem is that you can't have `field_validators` for `ClassVar`s??

Reply:
One thing that's happening here is you can't assign a new value to `settings_cls.eia860m_year_months` because the settings class is immutable.  That assignment by itself is causing an error:

```py
from pudl.settings import Eia860Settings
settings_cls = Eia860Settings()
settings_cls.eia860m_year_months = ["2019-01"]
```

I think what you're trying to test here is that if we attempt to construct `Eia860Settings` where both the EIA860 and the EIA860M dates include the same year, a `ValidationError` will be raised. I think that might look like:

```py
    def test_eia860_years_overlap_eia860m_years(self: Self):
        """Test validation error is raised when eia860m date is within eia860 years."""
        # Identify the last valid EIA-860 year:
        max_eia860_year = max(Eia860Settings().years)
        # Use that year to construct an EIA-860M year that overlaps the EIA-860 years:
        bad_eia860m_year_month = f"{max_eia860_year}-01"

        # Attempt to construct an EIA-860 settings object with an EIA-860M year that
        # overlaps the EIA-860 years, which should result in a ValidationError:
        with pytest.raises(ValidationError):
            _ = Eia860Settings(
                eia860m=True,
                years=[max_eia860_year],
                eia860m_year_months=[bad_eia860m_year_month],
            )
```

Reply:
This test as I've written it doesn't currently fail because of the missing return value that Christina highlighted in one of the validators. (`eia860m_year_months` is always `None` no matter what).  If you fix that return value, it still doesn't fail, and the settings object that's created looks like:

```py
Eia860Settings(
    disabled=False,
    years=[2022],
    eia860m=True,
    all_eia860m_year_months=[
        "2015-07",
        "2015-08",
        "2015-09",
        "2015-10",
        "2015-11",
        "2015-12",
        # etc. ...
        "2023-10",
        "2023-11",
        "2023-12",
        "2024-01",
        "2024-02",
        "2024-03",
    ],
    eia860m_year_months=[
        "2022-01",
        "2020-12",
        "2019-12",
        "2018-12",
        "2021-12",
        "2023-12",
        "2024-03",
        "2016-12",
        "2015-12",
        "2017-12",
    ],
)
```